promql/engine.go

```diff
@@ -705,11 +729,12 @@ func (ev *evaluator) rangeEval(f func([]Value, *EvalNodeHelper) Vector, exprs ..
 			vectors[i] = vectors[i][:0]
 			for si, series := range matrixes[i] {
 				for _, point := range series.Points {
-					if point.T == ts {
+					if point.T == ts && ev.currentSamples < ev.maxSamples {
 						vectors[i] = append(vectors[i], Sample{Metric: series.Metric, Point: point})
 						// Move input vectors forward so we don't have to re-scan the same
 						// past points at the next step.
 						matrixes[i][si].Points = series.Points[1:]
+						ev.currentSamples++
```

Comment:
You need to undo this before the next ts iteration

Reply:
Okay, so in that case I need a query for one of the tests that would load more than a single sample into memory per timestamp. The tests I have at the moment load at most one sample per ts, and I don't think setting the `ev.maxSamples` value to 0 is really a valid test.

Reply:
It's more testing across multiple timestamps

Reply:
How so? At the moment none of the tests I've added load more than a single sample into memory per ts.

Reply:
The issue is that you're still counting all the inputs from the previous timestamps when doing the current timestamp, but they're no longer in memory.

Reply:
Right, I've fixed that via lines 727/783. What I'm saying is that currently none of the tests have more than a single sample per timestamp. I want to find a query to add for a test that would load say 3 samples per timestamp into memory, and ensure the test fails if we set max samples to 2.